src/completion.ml

```diff
@@ -0,0 +1,292 @@
+and unif : unif_constrs -> unit = fun eqs ->
+  match eqs with
+  | []              -> ()
+  | (t1, t2) :: eqs -> unif_aux t1 t2 eqs
+
+(** [cps rs] computes the critical pairs of the rewrite system [rs] *)
+let cps : rule list -> (term * term) list = fun rs ->
+  let acc = ref [] in
+  let cps_aux : rule -> rule -> unit = fun r1 r2 ->
+    let lhs1, rhs1 = to_term r1 in
+    let lhs2, rhs2 = to_term r2 in
+    let find_cp : term -> term -> unit = fun t1 t2 ->
+      let reset_meta m = m.meta_value := None in
+      iter_meta reset_meta lhs1;
+      iter_meta reset_meta lhs2;
+      try
+        unif [(t1, t2)];
+        acc := (rhs1, rhs2) :: !acc
+      with Unsolvable -> () in
+    let _, args1 = get_args lhs1 in
+    let _, args2 = get_args lhs2 in
+    find_cp lhs1 lhs2;
+    List.iter (iter (find_cp lhs1)) args2;
+    List.iter (iter (fun t -> find_cp t lhs2)) args1 in
+  let rec cps rs =
+    match rs with
+    | []       -> ()
+    | r :: rs' -> List.iter (cps_aux r) rs; cps rs' in
+  cps rs;
+  !acc
+
+(** Rewriting for closed rules and closed terms *)
+
+type ctxt = term -> term
+
+(** [match_rule ctxt t (s, r)] attempts to rewrite [ctxt[t]] by appling the
+    rule [(s, t)]. *)
+let match_rule : ctxt -> term -> rule -> term option = fun ctxt t (s, r) ->
+  let lhs = add_args (symb s) r.lhs in
+  if eq t lhs then Some (ctxt (term_of_rhs r)) else None
+
+(** [match_rules ctxt t rs] attempts to rewrite [ctxt[t]] by appling the
+    rules [rs]. *)
+let match_rules : ctxt -> term -> rule list -> term option = fun ctxt t rs ->
+  List.map_find (match_rule ctxt t) rs
+
+(** [rewrite t rs] attempts to rewrite [t] in the rewrite system [rs]. *)
+let rewrite : term -> rule list -> term option = fun t rs ->
+  let t = unfold t in
+  match t with
+  | Wild
+  | TRef _
+  | Vari _
+  | Type
+  | Kind
+  | Patt _
+  | Meta _
+  | TEnv _
+  | Prod _
+  | Abst _      -> None
+  | Symb _      -> match_rules (fun t -> t) t rs
+  | Appl (t, u) ->
+      match match_rules (fun t -> t) t rs with
+      | None ->
+          let ctxt = fun t -> Appl (t, u) in
+          begin match match_rules ctxt t rs with
+          | None ->
+              let ctxt = fun u -> Appl (t, u) in
+              match_rules ctxt u rs
+          | res  -> res
+          end
+      | res  -> res
+
+(** [nf t rs] attempts to compute a normal form of [t] in the rewrite system
+    [rs]. *)
+let rec nf : term -> rule list -> term = fun t rs ->
+  match rewrite t rs with
+  | None   -> t
+  | Some t -> nf t rs
+
+(** [to_rule (lhs, rhs)] translates the pair [(lhs, rhs)] into a rule for
+    closed lhs and rhs. *)
+let to_rule : term * term -> rule = fun (lhs, rhs) ->
+  let (s, args) = get_args lhs in
+  let s = get_sym s in
+  let vs = Bindlib.new_mvar te_mkfree [||] in
+  let rhs = Bindlib.unbox (Bindlib.bind_mvar vs (lift rhs)) in
+  s, { lhs = args ; rhs = rhs ; arity = List.length args ; vars = [||] }
+
+(** [completion eqs ord] applies Huet's completion procedure on the set of
```

Comment:
Note that we do not need a so complex algorithm (with marked rules, etc.) because we only have closed equations.

Reply:
Note in particular that, since equations are closed, there is no need for unification.

Yet, unification is useful for checking the local confluence in the general case. That's why I propose to move unif and cps to a separate file dedicated to confluence.

Now, let's consider subject reduction. We start from a set of general unification constraints computed by type inference. We could proceed as follows:
1. Extract first-order equations from these constraints.
2. Replace metavariables by fresh function symbols with no type and no rule.
3. Orient the resulting equations by LPO. This gives a set R0 of rewriting rules. Update the rules of each symbol accordingly.
4. Here, we could compute critical pairs but, since equations are closed, we just need to do some normalization.
Assume for instance that we have a rule g->d and a rule l[g]->r. Then, we have a critical pair (l[d],r), but l[d] is just a reduct of l[g]. So, when we have a rule l->r, we could as well normalize l into l' with the rules distinct from l->r, and similarly normalize r into r'.
- If l'=r' then we can remove the rule l->r.
- If l=l' then there is no critical pair and thus nothing to do.
- If l<>l'>r' in LPO, then we can add the rule l'->r'.
- If l<>l'<r', then we can add the rule r'->l'.
Then, we start again with the new set of rules until there is no critical pair. Note that each new rule is obtained by reducing an initial equation by rules including in LPO. Therefore, this terminates.